### PR TITLE
WEBSITE-265: Generate redirects for old URLs

### PR DESCRIPTION
**Pull request by gunnarmorling**
This is largely done, so it's time to get some feedback. What has happened:

* _importer/importer.rb has been extended to create a .htaccess fragment with redirects for actual posts and tags
* _importer/user_redirect_creator.rb has been provided to generate a .htaccess fragment with redirects for user/blogger URLs
* _importer/crawler.rb has been extended to parse several URLs, with a depth of n (so to quickly crawl a given set of sub-pages)

That's the work-flow:

* Run the importer on the latest posts.pstore file (`bundle exec ./importer.rb -s ../posts-19-03-2015.pstore -o ../posts -rf ../redirects/.htaccess_posts -ni -na`): This will create the posts .htaccess fragment
* Create a pstore file with the home pages of all users (only once, could be shared via DropBox):
    bundle exec ./crawler.rb -u "http://in.relation.to/Bloggers/Ales,http://in.relation.to/Bloggers/Aslak,http://in.relation.to/Bloggers/Brett,http://in.relation.to/Bloggers/Christian,http://in.relation.to/Bloggers/Dan,http://in.relation.to/Bloggers/Daniel,http://in.relation.to/Bloggers/David,http://in.relation.to/Bloggers/Davide,http://in.relation.to/Bloggers/Emmanuel,http://in.relation.to/Bloggers/Gail,http://in.relation.to/Bloggers/Gavin,http://in.relation.to/Bloggers/Gleb,http://in.relation.to/Bloggers/Greg,http://in.relation.to/Bloggers/Gunnar,http://in.relation.to/Bloggers/Hardy,http://in.relation.to/Bloggers/Ilya,http://in.relation.to/Bloggers/Jason,http://in.relation.to/Bloggers/JasonP,http://in.relation.to/Bloggers/Jay,http://in.relation.to/Bloggers/Jesper,http://in.relation.to/Bloggers/John,http://in.relation.to/Bloggers/Jozef,http://in.relation.to/Bloggers/Ken,http://in.relation.to/Bloggers/Lincoln,http://in.relation.to/Bloggers/Luksa,http://in.relation.to/Bloggers/Marek,http://in.relation.to/Bloggers/Marius,http://in.relation.to/Bloggers/Marko,http://in.relation.to/Bloggers/Martin,http://in.relation.to/Bloggers/Max,http://in.relation.to/Bloggers/Michael,http://in.relation.to/Bloggers/Nicklas,http://in.relation.to/Bloggers/Norman,http://in.relation.to/Bloggers/Pete,http://in.relation.to/Bloggers/Robb,http://in.relation.to/Bloggers/Sanne,http://in.relation.to/Bloggers/Scott,http://in.relation.to/Bloggers/Sergey,http://in.relation.to/Bloggers/Shane,http://in.relation.to/Bloggers/Snjezana,http://in.relation.to/Bloggers/Steve,http://in.relation.to/Bloggers/StrongLiu,http://in.relation.to/Bloggers/Stuart,http://in.relation.to/Bloggers/Tihomir" -p ".*" -o users.pstore -d 0
* Create the blogger .htaccess fragement: `bundle exec ./user_redirect_creator.rb -s users.pstore -o ../redirects/.htaccess_bloggers`
* Merge the fragments to a combined .htaccess file: `cat ../redirects/.htaccess_misc ../redirects/.htaccess_bloggers ../redirects/.htaccess_posts > ../_site/.htaccess`

Open questions:

* Are all (important) URLs covered? What is missing (manual additions can go into .htaccess_misc)?
* Casing of posts; I saw Emmanuel and Hardy discuss about this yesterday. If casing differs between old and new URLs, that'd have to be considered
* What to do about bloggers with 0 posts? E.g. http://in.relation.to/Bloggers/Snjezana has no posts at all; I think we can ignore those from the beginning?
* Is there a user profile page already? I.e. a counter-part to e.g. http://in.relation.to/user/gunnar, or should these just redirect to the user's posts?
* Instead of .htaccess the redirects should go into the main Apache config for perf reasons; .htaccess is quicker for development though as changes get instantly visible
* Currently headers are sent with 301 responses to suppress browser caching of redirects; That's good for now as otherwise it's nearly impossible to fix wrong re-directs (users would have to empty browser caches); But in the long run, if the redirects prove good, we could remove these headers and enable caching
* Just coming to my mind, what about attachments? Is it worth to create redirects for those as well?
* Need to incorporate tag aggregation rules created as part of WEBSITE-299

### CODE REVIEW

**Comment by gunnarmorling** [comment]
@Sanne I've assigned this to you as you seemed interested. Any feedback welcome.

**Comment by Sanne** [comment]
> Currently headers are sent with 301 responses to suppress browser caching of redirects; That's good for now as otherwise it's nearly impossible to fix wrong re-directs (users would have to empty browser caches); But in the long run, if the redirects prove good, we could remove these headers and enable caching

? the http spec defines 301 as cacheable. And since it's "redirect-permanent" it makes sense right?

**Comment by Sanne** [comment]
> Just coming to my mind, what about attachments? Is it worth to create redirects for those as well?

Assuming our blog posts still link to them correctly (we fixed our internal links right?), I'd say that's not necessary.

**Comment by Sanne** [comment]
I did run

 > bundle install

But still I keep getting:

 > bundle exec ./importer.rb -s ../posts-19-03-2015.pstore -o ../posts -rf ../redirects/.htaccess_posts -ni -na
bundler: command not found: ./importer.rb
Install missing gem executables with `bundle install`

**Comment by gunnarmorling** [comment]
> the http spec defines 301 as cacheable. And since it's "redirect-permanent" it makes sense right?

Yes, of course it makes sense. Only we need to be sure that the re-directs are like we want them.

E.g. if we start with a single feed URL (and re-direct all old personalized feed URLs to that) and later on come to the idea that actually a feed per blogger would be nice, it'll be a problem to change the personal feed re-directs to point to the new personalized feeds from there-on. Hence my approach for disallowing clients to cache the redirects during some stabilization phase. At least while we are actively working on it, otherwise it's a pain to even test it.

**Comment by gunnarmorling** [comment]
> But still I keep getting ...

To be sure, are you running this from within the "_importer" directory? Maybe you are running it from the root dir instead?

File: redirects/.htaccess_bloggers [hunk 1]
```diff
@@ -0,0 +1,432 @@
+# Generated by user_redirect_creator.rb
+
+# Ales
+RewriteRule ^Bloggers/Ales$ /author/ales-justin/ [R=301,L,E=nocache:1]
+RewriteRule ^Bloggers/AlesHome$ /author/ales-justin/ [R=301,L,E=nocache:1]
+RewriteRule ^Bloggers/AlesHome/Page/.*$ /author/ales-justin/ [R=301,L,E=nocache:1]
+RewriteCond %{REQUEST_URI} ^/dirDisplay_d.seam$
```
Comment (hferentschik) [inline redirects/.htaccess_bloggers]: what are these about?

File: redirects/.htaccess_bloggers [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+RewriteCond %{QUERY_STRING} ^directoryId=9762$
+RewriteRule ^(.*)$ /author/ales-justin/? [R=301,L,E=nocache:1]
```
Comment (hferentschik) [inline redirects/.htaccess_bloggers]: I guess this works somehow in combination with RewriteCond?

File: _importer/importer.rb [hunk 3]
```diff
@@ -244,11 +283,11 @@ def prepare_content_body(doc, blog_entry)
       end
     end
 
-    unless @skip_image_procesing
+    unless @skip_image_processing
       import_images(content_node)
     end
 
-    unless @skip_asset_procesing
+    unless @skip_asset_processing
```
Comment (hferentschik) [inline _importer/importer.rb]: :-)

**Comment by hferentschik** [comment]
The posts re-direct file is still empty, right? I assume this is still work in progress? Overall all the modifications look ok. 

Some  comments to your questions:

> Casing of posts; I saw Emmanuel and Hardy discuss about this yesterday. If casing differs between old and new URLs, that'd have to be considered

Right, that's WEBSITE-327. I have some code in place for that and posted how the slugs would look like. Waiting for some feedback. If we want to go down this road, I'd just need to create a pull request.

> What to do about bloggers with 0 posts? E.g. http://in.relation.to/Bloggers/Snjezana has no posts at all; I think we can ignore those from the beginning?

I think we can safely remove them

> Is there a user profile page already? I.e. a counter-part to e.g. http://in.relation.to/user/gunnar, or should these just redirect to the user's posts?

We need to create them (WEBSITE-276). They go into __data/people/_ and there are three examples already. I guess we can do this manually once off.

> Instead of .htaccess the redirects should go into the main Apache config for perf reasons; .htaccess is quicker for development though as changes get instantly visible

+1 - even though I'd prefer to site dedicated config file. On several Linux distributions that is already pre-configured and once just needs to create a in.relation.to.conf in the right directory. 

> Just coming to my mind, what about attachments? Is it worth to create redirects for those as well?

-1 I don't think this is necessary.

**Comment by emmanuelbernard** [comment]
Gunnar can you share the .htaccess or .htaccess files, with my super slow connection it is not practical to  run the importer

**Comment by emmanuelbernard** [comment]
> Create a pstore file with the home pages of all users (only once, could be shared via DropBox): bundle exec ./crawler.rb -u "http://in.relation.to/Bloggers/Ales,http://in.relation.to/Bloggers/Aslak,http://in.relation.to/Bloggers/Brett,http://in.relation.to/Bloggers/Christian,http://in.relation.to/Bloggers/Dan,http://in.relation.to/Bloggers/Daniel,http://in.relation.to/Bloggers/David,http://in.relation.to/Bloggers/Davide,http://in.relation.to/Bloggers/Emmanuel,http://in.relation.to/Bloggers/Gail,http://in.relation.to/Bloggers/Gavin,http://in.relation.to/Bloggers/Gleb,http://in.relation.to/Bloggers/Greg,http://in.relation.to/Bloggers/Gunnar,http://in.relation.to/Bloggers/Hardy,http://in.relation.to/Bloggers/Ilya,http://in.relation.to/Bloggers/Jason,http://in.relation.to/Bloggers/JasonP,http://in.relation.to/Bloggers/Jay,http://in.relation.to/Bloggers/Jesper,http://in.relation.to/Bloggers/John,http://in.relation.to/Bloggers/Jozef,http://in.relation.to/Bloggers/Ken,http://in.relation.to/Bloggers/Lincoln,http://in.relation.to/Bloggers/Luksa,http://in.relation.to/Bloggers/Marek,http://in.relation.to/Bloggers/Marius,http://in.relation.to/Bloggers/Marko,http://in.relation.to/Bloggers/Martin,http://in.relation.to/Bloggers/Max,http://in.relation.to/Bloggers/Michael,http://in.relation.to/Bloggers/Nicklas,http://in.relation.to/Bloggers/Norman,http://in.relation.to/Bloggers/Pete,http://in.relation.to/Bloggers/Robb,http://in.relation.to/Bloggers/Sanne,http://in.relation.to/Bloggers/Scott,http://in.relation.to/Bloggers/Sergey,http://in.relation.to/Bloggers/Shane,http://in.relation.to/Bloggers/Snjezana,http://in.relation.to/Bloggers/Steve,http://in.relation.to/Bloggers/StrongLiu,http://in.relation.to/Bloggers/Stuart,http://in.relation.to/Bloggers/Tihomir" -p ".*" -o users.pstore -d 0

Looks like hardcoding the 30 or so users would have been faster.

> Are all (important) URLs covered? What is missing (manual additions can go into .htaccess_misc)?

See my previous comment, the generated htaccess files would help me.

> What to do about bloggers with 0 posts? E.g. http://in.relation.to/Bloggers/Snjezana has no posts at all; I think we can ignore those from the beginning?

Yes ignore them.

> Is there a user profile page already? I.e. a counter-part to e.g. http://in.relation.to/user/gunnar, or should these just redirect to the user's posts?

`/author/emmanuel-bernard` (or soon `/emmanuel-bernard`) has the bio on the left side. We can add a pic and that would probably suffice as a good enough landing page.

> Instead of .htaccess the redirects should go into the main Apache config for perf reasons; .htaccess is quicker for development though as changes get instantly visible

Is it really bad? I find it very useful to have the .htaccess rules in the SCM and updatable by a single commit.

> Currently headers are sent with 301 responses to suppress browser caching of redirects; That's good for now as otherwise it's nearly impossible to fix wrong re-directs (users would have to empty browser caches); But in the long run, if the redirects prove good, we could remove these headers and enable caching

you can simply use 302 for the mean time.

> Just coming to my mind, what about attachments? Is it worth to create redirects for those as well?

Nope

**Comment by Sanne** [comment]
I was also about to suggest using 302 for the moment. I think 301 states that it's a done deal, even if you *suggest* to clients to not cache ...

File: redirects/.htaccess_bloggers (hunk 1 above)
Comment (gunnarmorling) [inline redirects/.htaccess_bloggers]: It's the list of a blogger's posts, if you click on the "Browse" button in the upper right of your home page, e.g. at http://in.relation.to/Bloggers/Hardy.

File: redirects/.htaccess_bloggers (hunk 2 above)
Comment (gunnarmorling) [inline redirects/.htaccess_bloggers]: Yes, the two `RewriteCond` are applied by the subsequent `RewriteRule`.

**Comment by gunnarmorling** [comment]
> The posts re-direct file is still empty, right?
> Gunnar can you share the .htaccess or .htaccess files

The files are all under redirects as part of this PR (the contents of .htaccess_posts is only not shown in the GitHub UI as its too large):

* redirects/.htaccess_bloggers
* redirects/.htaccess_misc
* redirects/.htaccess_posts

**Comment by gunnarmorling** [comment]
> We need to create them (WEBSITE-276). They go into __data/people/_ and there are three examples already.

Ah, ok. At which URL are these exposed?

> +1 - even though I'd prefer to site dedicated config file. On several Linux distributions that is already pre-configured and once just needs to create a in.relation.to.conf in the right directory. 

We can make it it a separate file and include it into the main config if we like. It only should not be a .htaccess as these get re-evaluated for each request.

**Comment by gunnarmorling** [comment]
> I seriously doubt that this piece of code was faster to write than using the list of static bloggers and write the htaccess rules in the misc file :)

That way we can re-run it and adapt as needed. I actually started manually as you suggest but it was just too cumbersome to update it with reg-exs etc. Also it's a matter of avoiding errors done in manual processing.

**Comment by gunnarmorling** [comment]
> Looks like hardcoding the 30 or so users would have been faster.

Check out the users file, it's more than a single re-direct per user. It helped me to script it to get it right.

> Is it really bad? I find it very useful to have the .htaccess rules in the SCM and updatable by a single commit.

The files are in SCM (see the "redirects" directory). Only it should be includes into the main config rather than a .htaccess file for perf reasons.

> you can simply use 302 for the mean time.

You mean prior to the go-live? Yes, but if we change to 301 when going live and then detecting the need for changes it gets troublesome.

>I think 301 states that it's a done deal, even if you suggest to clients to not cache ...

I see it's not cached in Firefox and Chrome, I'd expect the same for other browsers.

**Comment by gunnarmorling** [comment]
> the generated htaccess files would help me.

The bloggers part is at https://github.com/hibernate/in.relation.to/pull/19/files#diff-af585a5e5d64c69ed545a65660909e2b. You need to fetch the PR to see the posts part, as it's too large to be shown on the GitHub UI.

**Comment by gunnarmorling** [comment]
Rebased and added two more commits to accommodate to the new post URL casing and to remove bloggers with 0 posts.

**Comment by gunnarmorling** [comment]
Wrt. 302 redirects, I thought we don't want to use them because the loss of page rank? Note that I mean to disable caching of redirects for a while *after* go-live so we can fix any wrong redirects. Also see [this post](http://getluky.net/2010/12/14/301-redirects-cannot-be-undon/) for some background; as said, 301 redirects cannot be altered if caching hasn't been suppressed, hence I'd only do this after some grace period after go-live.

**Comment by hferentschik** [comment]
Where do we stand on this?

**Comment by gunnarmorling** [comment]
It's good to go from my side. I only need to adapt to the altered tags, but I can do that in a separate PR.

**Comment by hferentschik** [comment]
I merged the changes to the tags, so maybe you could re-base and adjust. Then we can merge it to master as a first cut. Chances are we need to re-run this anyways a couple of times.

**Comment by Sanne** [comment]
right, looks good and +1 to use those strange non-cacheable 301 (I was mostly surprised about it)

Note that if you use 302 you lose SEO value but temporarily, only until you fix it.. just saying, I don't feel strong about either way.

The bigger question: can we get the httpd configuration files generated in the right place, or is that better done manually at the end?
I'm not sure I've understood when the file-generation is going to be invoked.

**Comment by gunnarmorling** [comment]
Ok, I will work out and push one more commit to adapt to the tag normalization.

> Chances are we need to re-run this anyways a couple of times.

Yes, that's what I am expecting, too. As it's all done in scripts that's no problem.

> can we get the httpd configuration files generated in the right place, or is that better done manually at the end?

We could generate the final redirects file as part of the build, but I don't think it's strictly needed. It's basically just cat-ing the fragments, and moving it to the right folder of Apache which I guess needs to be done manually anyways (assuming we include it into the main config and not via a .htaccess in the _site dir)

> I'm not sure I've understood when the file-generation is going to be invoked.

The posts and blogger fragements are created upon importing the contents of the old blog (i.e. when running importer.rb, and now new, create-user-redirects.rb). The misc fragment is hand-written and holds any additional redirects we deem necessary. The merge of the fragments happens upon site creation.

**Comment by Sanne** [comment]
> It's basically just cat-ing the fragments, and moving it to the right folder of Apache which I guess needs to be done manually anyways (assuming we include it into the main config and not via a .htaccess in the _site dir)

I guess we need to decide on that. I'm totally neutral about where we keep it, in both cases it's going to be in a git repository.
Having it within the site directory seems a bit easier since not everyone has Ansible setup like we have awstruct, I'd start from there and move it only if we need it.

**Comment by gunnarmorling** [comment]
> in both cases it's going to be in a git repository.

We don't need the resulting redirects file in Git, we have the fragments which make it up (under "redirects").

> Having it within the site directory seems a bit easier

Under _site (which is the output folder of the generated website) it only really makes sense if we want to go for .htaccess in the end. Which I think we should not (for perf reasons). Rather we should include this into the main Apache config.

What I can do for development/testing purposes for now is to augment the "gen" task to emit a .htaccess file under "_site", but as said that's not the way I think we should pursue for prod in the end.

**Comment by Sanne** [comment]
I'm not sure if we should bother on the performance aspect, but ok.
No need to make a .htaccess file then, I'll feed the current fragments to the apache global config for testing.

**Comment by gunnarmorling** [comment]
> I'm not sure if we should bother on the performance aspect, but ok.

Wow, that I hear these words out of your mouth :) I will mark the day in the calendar ;)

**Comment by Sanne** [comment]
lol :)
It's just that I don't expect such an high load and the machine is quite high spec. It's good to know about such things so that we can enable in case of need.. but the .htaccess documentation seems to point out server usage, not end-user perceived usage which I would fight hard for!

**Comment by hferentschik** [comment]
> Wow, that I hear these words out of your mouth :) I will mark the day in the calendar ;)

I made a cross in mine as well ;-)

**Comment by gunnarmorling** [comment]
Pushed two more commits to add some more tag mappings and consider the mappings during re-direct creation.

**Comment by emmanuelbernard** [comment]
A few questions @gunnarmorling 
- is this redirect/ directory excluded from the _site (awestruct_ignore I think)?
- how can I add redirects manually later during the life of the project? As part of misc and it will always reconfigure httpd and stop restart it for every build?

**Comment by emmanuelbernard** [comment]
```
# Ales
RewriteRule ^Bloggers/Ales$ /ales-justin/ [R=301,L,E=nocache:1]
RewriteRule ^Bloggers/AlesHome$ /ales-justin/ [R=301,L,E=nocache:1]
RewriteRule ^Bloggers/AlesHome/Page/.*$ /ales-justin/ [R=301,L,E=nocache:1]
RewriteCond %{REQUEST_URI} ^/dirDisplay_d.seam$
RewriteCond %{QUERY_STRING} ^directoryId=9762$
RewriteRule ^(.*)$ /ales-justin/? [R=301,L,E=nocache:1]
RewriteRule ^user/alesj$ /ales-justin/ [R=301,L,E=nocache:1]
RewriteRule ^service/Feed/atom/Area/Bloggers/Node/Ales(/Comments/.*)? /blog.atom [R=301,L,E=nocache:1]

# Aslak
RewriteRule ^Bloggers/Aslak$ /aslak-knutsen/ [R=301,L,E=nocache:1]
RewriteRule ^Bloggers/AslaksBlog$ /aslak-knutsen/ [R=301,L,E=nocache:1]
RewriteRule ^Bloggers/AslaksBlog/Page/.*$ /aslak-knutsen/ [R=301,L,E=nocache:1]
RewriteCond %{REQUEST_URI} ^/dirDisplay_d.seam$
RewriteCond %{QUERY_STRING} ^directoryId=13996$
RewriteRule ^(.*)$ /aslak-knutsen/? [R=301,L,E=nocache:1]
RewriteRule ^user/aslak$ /aslak-knutsen/ [R=301,L,E=nocache:1]
RewriteRule ^service/Feed/atom/Area/Bloggers/Node/Aslak(/Comments/.*)? /blog.atom [R=301,L,E=nocache:1]

```

I am not super familiar with htaccess but doesn't it imply that the first rules of Aslak are only executed when the `RewriteCond` of Ales are true?

**Comment by emmanuelbernard** [comment]
Tag consolidation is not taken into account yet right?

**Comment by emmanuelbernard** [comment]
I don't see any obvious missing URLs but I will keep poking around.

**Comment by gunnarmorling** [comment]
> Tag consolidation is not taken into account yet right?

Yes, it is. Check out the end of the .htaccess_posts file.

**Comment by hferentschik** [comment]
I actually merged this now. As mentioned, we will need to do re-run anyways and probably do more adjustments. Also I want to save @gunnarmorling from even more re-basing and merging, if I make changes for the comment imports.

**Comment by gunnarmorling** [comment]
> doesn't it imply that the first rules of Aslak are only executed when the RewriteCond of Ales are true?

No, RewriteConds only affect the next RewriteRule following.

**Comment by gunnarmorling** [comment]
> is this redirect/ directory excluded from the _site (awestruct_ignore I think)?

I haven't excluded it explicitly, but it's not in _site for me. I'll see whether I can exclude it to be safe.

> how can I add redirects manually later during the life of the project? As part of misc and it will always reconfigure httpd and stop restart it for every build?

Yes, as part of misc. The exact way for making it active then depends on the route we take: .htaccess changes take effect immediately (as it's re-examined for each access). The main config needs to reloaded explicitly, but that can be done without a restart (could e.g. happen as part of the deploy job).

**Comment by gunnarmorling** [comment]
> I actually merged this now. 

Sweet, thanks.

**Pull request merged**